qiita_pet/handlers/analysis_handlers/base_handlers.py

```diff
@@ -125,36 +126,17 @@ def analyisis_graph_handler_get_request(analysis_id, user):
     full_access = (analysis in (user.private_analyses | user.shared_analyses)
                    or user.level in {'superuser', 'admin'})
 
-    nodes = set()
-    edges = set()
+    nodes = []
+    edges = []
+    wf_id = None
     # Loop through all the initial artifacts of the analysis
     for a in analysis.artifacts:
-        g = a.descendants_with_jobs
-        # Loop through all the nodes in artifact descendants graph
-        for n in g.nodes():
-            # Get if the object is an artifact or a job
-            obj_type = n[0]
-            # Get the actual object
-            obj = n[1]
-            if obj_type == 'job':
-                otype = 'job'
-                name = obj.command.name
-            elif not full_access and not obj.visibility == 'public':
-                # The object is an artifact, it is not public and the user
-                # doesn't have full access, so we don't include it in the
-                # graph
-                continue
-            else:
-                otype = obj.artifact_type
-                name = '%s\n(%s)' % (obj.name, otype)
-            nodes.add((obj_type, otype, obj.id, name))
-
-        edges.update({(s[1].id, t[1].id) for s, t in g.edges()})
-
-    # Nodes and Edges are sets, but the set object can't be serialized using
-    # JSON. Transforming them to lists so when this is returned to the GUI
-    # over HTTP can be JSONized.
-    return {'edges': list(edges), 'nodes': list(nodes)}
+        if a.processing_parameters is None:
+            g = a.descendants_with_jobs
```

Comment:
My only concern with this approach is that you are generating the initial graph `a.descendants_with_jobs` and then formatting to JS `get_network_nodes_edges`. In other words, you are doing the same thing twice. What about generating the JS from within the artifact? This might safe loops and we can think it as a different `print` format for the artifact ...

Reply:
That would be coupling the representation of the data to the current interface. Note that `descendants_with_jobs` is returning a rich object (network.DiGraph) which we use in other parts of the code to access those objects. However, the JS representation of it is just 2 lists (nodes and edges) which will make it harder to do any traversing of the graph (i.e. we would need to re-implement those traversals that are already in networkx).

Reply:
After offline discussion, we will see general behavior on test server and reevaluate.